core/validatorapi/grpc/validator.go

```diff
@@ -0,0 +1,512 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+
+	eth2client "github.com/attestantio/go-eth2-client"
+	eth2api "github.com/attestantio/go-eth2-client/api"
+	eth2v1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/attestantio/go-eth2-client/spec"
+	"github.com/attestantio/go-eth2-client/spec/altair"
+	"github.com/attestantio/go-eth2-client/spec/bellatrix"
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/jonboulle/clockwork"
+	pb "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	prysmpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+
+	"github.com/obolnetwork/charon/app/eth2wrap"
+	"github.com/obolnetwork/charon/core"
+	"github.com/obolnetwork/charon/core/scheduler"
+)
+
+var _ prysmpb.BeaconNodeValidatorServer = (*BeaconNodeValidatorServer)(nil)
+
+// BeaconNodeValidatorServer implements the server API for BeaconNodeValidator grpc service.
+//
+// The beacon node validator API enables a validator to connect
+// and perform its obligations on the Ethereum Beacon Chain.
+type BeaconNodeValidatorServer struct {
+	dutyDB             core.DutyDB
+	syncDutiesProvider eth2client.SyncCommitteeDutiesProvider
+	attDutiesProvider  eth2client.AttesterDutiesProvider
+	propDutiesProvider eth2client.ProposerDutiesProvider
+	prepareSubmitter   eth2client.ProposalPreparationsSubmitter
+	valsProvider       eth2client.ValidatorsProvider
+	eventsProvider     eth2client.EventsProvider
+	blockProvider      eth2client.SignedBeaconBlockProvider
+
+	eth2Cl     eth2wrap.Client
+	builderAPI bool
+
+	mu      sync.Mutex
+	streams map[int64]func(epoch eth2p0.Epoch)
```

Comment:
// map[slot][epoch] ??

Reply:
these are callbacks that take an epoch as param